### ISSUE TITLE
Split on Argument in Anonymous Functions should rewrite to Pattern Match syntax

### ISSUE THREAD

**Issue by marco-vassena** [agda/agda#2777]
Agda version 2.5.3. 
In the interactive agda-mode on emacs , case split on the argument of an anonymous function fails.

Consider the following:
```agda
data Bool : Set where
  true : Bool
  false : Bool

not : Bool → Bool
not = λ x → {!!}
```

Splitting on x (C-c C-c x RET), will produce the following error:
```
Cannot split on local variable x
when checking that the expression ? has type Bool
```

However, if curly brackets are present it succeeds.
That is C-c C-c x RET on the following:
```agda
not : Bool → Bool
not = λ { x → {!!} }
```
rewrites the function as
```agda
not : Bool → Bool
not = λ { true → {!!} ; false → {!!} }
```
as expected.

I believe that C-c C-c x RET should give the latter, regardless of the explicit presence of curly brackets.

**Comment by jespercockx** [comment]
Currently, there is a semantic difference between regular and extended lambdas (i.e. without and with curly braces): regular lambdas are compared structurally, but each extended lambda introduces a new (anonymous) top-level function. So we have `λ x → x == λ x → x`, but `λ { x → x } != λ { x → x }`. Since they have a different semantics, I think it would be misleading to turn the former into the latter automatically during case splitting.

**Comment by marco-vassena** [comment]
I see. I just started using pattern match in anonymous lambdas (which I find super-useful to deal with short proofs by contradiction), and I did not realise that they have different semantics. It seems to be a rather subtle difference.

As a user, I would argue that if I do case split (an explicit `C-c C-c x RET` command), then my intention is clearly to use the extended lambda semantics instead of the regular lambda.

As a matter of fact, the workflow after the error message is to manually rewrite `λ x → {!!}` to `λ { x → {!!} }` and then do case split (note that `λ x → {!!}` is the lambda that you get after refine, i.e., `C-c C-r`).

From the user experience prospective, the suggestion that I propose would make proofs that follow this pattern faster to complete: it does not require any manual editing, just a sequence of refine (`C-c C-r`) and split (`C-c C-c x RET`) commands will do.

**Comment by asr** [comment]
@marco-vassena, please note that since you replied via email, we cannot distinguish your comments from @jespercockx comments.

**Comment by UlfNorell** [comment]
I agree with Marco here. If the user asks for a case split, I don't see how it could possibly be misleading to perform one.

**Comment by andreasabel** [comment]
What would you expect in case there are several lambdas?
```
and : (x y : Bool) -> Bool
and = \ x y -> {! x !} -- C-c C-c

or : (x y : Bool) -> Bool
or = \ x -> \ y -> {! x !} -- C-c C-c
```
Maybe
```
and = \{ true y -> ?; false y -> ? }

or = \{ true -> \ y -> ?; false -> \ y -> ? }
```
Or should it not matter whether the lambdas are joined or not?  

Internally, both are represented the same:
```agda
postulate
  A : Set
  P : (A → A → A) → Set

test1 : P λ x y → x
test1 = {!!}

test2 : P λ x → λ y → x
test2 = {!!}
```
prints the goals the same:
```
?0 : P (λ x y → x)
?1 : P (λ x y → x)
```
Thus, without extra infrastructure we would also not be able to distinguish them for case splitting.

**Comment by andreasabel** [comment]
This proposal needs to be fleshed out more before implementation can start. The behavior for cases involving several variables needs to be specified, including: several, lambdas, several variables in a lambda, several variables to split at the same time.

**Comment by UlfNorell** [comment]
I think it's a mistake to spend too much time on the specification before starting implementation. Without at least starting the implementation it will be hard to figure out what the limiting factors are. For instance, what can reliably be achieved on the Emacs side.

Also, I don't see a problem in implementing something that works for the common simple case, but fails (gracefully) in more complex scenarios. At the moment all scenarios fails, so it's shouldn't be hard to improve on.

**Comment by andreasabel** [comment]
For a start, my questions raised above need to be answered.  They arose when I thought about a possible implementation.

Most of the work would have to be done already for a simpler feature:
==> Turn an ordinary lambda into a pattern-lambda.

Likely, concerning emacs interaction, we will hit similar problems as we have now for case-splitting. https://github.com/agda/agda/labels/case-splitting

**Comment by marco-vassena** [comment]
I don't think that it should matter whether the lambdas are joined or not and I would expect case splitting to retain which ever style the user used.

**Comment by andreasabel** [comment]
> I would expect case splitting to retain which ever style the user used.

I would so, too, but this requires first some refactoring to actually remember how the user wrote the lambdas.

**Comment by UlfNorell** [comment]
I think it wouldn't be the end of the would if split lambdas were lost. I would propose implementing this feature in two steps: first without worrying about preserving split lambdas, and second tracking more information about how the user wrote their lambdas.
